packages/std/src/types.rs

```diff
@@ -178,6 +178,29 @@ impl QueryResult {
     }
 }
 
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum QueryRequest {
+    // this queries the public API of another contract at a known address (with known ABI)
+    // msg is the json-encoded QueryMsg struct
+    // return value is whatever the contract returns (caller should know)
+    Contract {
+        contract_addr: HumanAddr,
+        msg: Binary, // we pass this in as Vec<u8> to the contract, so allow any binary encoding (later, limit to rawjson?)
```

Comment:
Couldn't we use `String` as a better JSON encoding? It ensures valid UTF-8, is human readable and does not have the 1/3 encoding overhead of base64.

Reply:
A few reasons.

First our Json serializer doesn't escape strings. Thus if you have `"` inside the message body (eg. Any non empty Json message) it will break. This is why I moved to vec  (and then base64). If we really want json with minimal overhead, we would use a raw json type that accepts a valid json object here (not string encoded)

Also, I am still unsure how much we want to commit to json long term for encoding and have minimal hard dependencies (now anything supported by serde) to allow changing it later. We can revisit this decision in 0.8

3rd I wanted to keep consistency between the value passed here to the contract and what the contract receives in the query method. They should be the same type with no processing by the runtime.

Reply:
Interesting, thanks